### PR TITLE
[SATURN-1892] Uploader bugfixes and performance improvements

### PR DESCRIPTION
**Pull request by andrewmarcus**
## Apr 2 update

* I switched the reload buffering in the FileBrowser to 5 seconds (rather than 10), to be consistent with the comment and also because it provides the user with slight more real-time feedback. This still only matters when uploading lots of small files; if a big file takes longer than 5 seconds, it will still only refresh once the file is fully uploaded.
* I switched the FileBrowser component to use a `FlexTable` rather than a `GridTable`, which is better suited for this use case.
* Both `FlexTable` and `GridTable` now accept optional parameters to change the row heights and override the header and cell styles. This allowed me to condense the rows and remove borders, to get the file browser to more closely match the way it looks in the Data tab in Workspaces. FWIW, that one should probably be updated to use a FlexTable too, or probably just reuse the new FileBrowser component, but I figured it would be better not to make that change in this PR.
* I updated the way the Metadata preview table is rendered to show more clearly which columns will be updated versus added, when updating an existing table. This provides a better user experience while dedicating more screen real-estate to the table, better supporting smaller displays.

## Original

This PR switches the table of uploaded files from a `SimpleTable` to a `GridTable`, so that only the visible rows are rendered into the DOM. This appears to resolve the performance issues Joel encountered when labs uploaded hundreds of thousands of files.

In order to make this look good, I added some new styling options to GridTable. You can now set the desired height of each header row and data row, as well as apply additional styling options to the header. This makes the file table in the uploader look much more like the one in the Data tab within Workspaces.  I was unable to get the delete button to hide except on hover, but this does not seem to be important enough to spend time on.

To further improve performance, I reduced the frequency with which the file table is update while files are uploading. Previously, every time a file finished the table would refresh so that users can see the new file. However, this is rather meaningless when uploading hundreds of files at a time since the table only renders a few of them, and results in unnecessary work. I've debounced it so the table will refresh no more often than once every 5 seconds.

In addition, I fixed the bug which was causing only the first thousand filenames in a metadata file to be translated to their bucket URLs. We will now grab as many pages of 1000 URLs as we can before attempting to do the matching.

On the other, the one caveat of the file table is that it only displays the first 1000 files in the bucket. If it's important, I can figure out how to add a pager, or do automatic paging like I do to count the URLs. But I'm not sure whether it's necessary for most users.

### Metadata Uploader improvements

To improve the user experience, I reduced the vertical size of the Metadata Upload panel so that more of the metadata table is visible.  Further improvements could be made here, such as listing the columns to be updated in a row rather than a bulleted list. But this is better than it was.

I also fixed a bug preventing table columns from being sorted properly. This is working now.

### CODE REVIEW

File: src/components/ProgressBar.js
Comment (andrewmarcus) [inline src/components/ProgressBar.js:5]: Turns out I wrote an unnecessary function, so I'm removing it and falling back on the one which already exists and is used everywhere else

File: src/components/data/FileBrowser.js
Comment (panentheos) [inline src/components/data/FileBrowser.js]: Just a preliminary note: If this table isn't intended to scroll horizontally, the `FlexTable` component might be a better fit, since it will give you virtualized rows and you can specify flexbox-like constraints for the column sizes.

Comment (andrewmarcus) [reply src/components/data/FileBrowser.js]: Ah, thanks!  Yeah, I saw that one but hadn't looked much into it.

Happy to switch to that one before this PR gets merged, or come back and fix it later. Whichever you'd prefer...

File: src/libs/ajax.js
Comment (breilly2) [inline src/libs/ajax.js]: Nit: May as well keep using `_.concat` here for consistency.

File: src/components/data/UploadPreviewTable.js
Comment (breilly2) [inline src/components/data/UploadPreviewTable.js:113]: Nit: `_.findIndex` should work fine here, again for consistency.

File: src/components/data/FileBrowser.js
Comment (breilly2) [inline src/components/data/FileBrowser.js:91]: In the PR description (and the code below), you used 10 seconds. IMO, 5 is probably enough of a rate limit, and provides more immediate feedback, but as long as the code and this comment are consistent, it doesn't really matter to me.

File: src/components/data/FileBrowser.js
Comment (breilly2) [inline src/components/data/FileBrowser.js]: TIL about using exponents in numeric literals!

File: src/components/data/FileBrowser.js
Comment (breilly2) [reply src/components/data/FileBrowser.js]: Andrew, I'm going to leave it up to you. As long as the current code meets your need, I think it's fine to go ahead with it and refactor later.

**Review by breilly2 — APPROVED** [review]
👍  Looks good. A couple of comments, but nothing blocking.

I mostly studied the shared code. I looked over the uploader code, but I'm not familiar with the feature and didn't try using it. Andrew, I'll trust whatever testing you did yourself. Let me know if you want more attention to testing in this review.

I did run the integration tests locally and they passed.

File: src/components/data/FileBrowser.js
Comment (andrewmarcus) [reply src/components/data/FileBrowser.js]: I just pushed an update to switch to a `FlexTable`. You're right, the conversion was trivial and it's definitely a cleaner implementation for this use case.

I added `rowHeight` and `headerHeight` parameters, as well as `styleCell()` and `styleHeader()` functions, in order to attempt to match the styling of File Browsers elsewhere on the site. Since these are all optional, their presence should be a no-op for anything else using these tables.

File: src/components/data/UploadPreviewTable.js
Comment (andrewmarcus) [reply src/components/data/UploadPreviewTable.js:113]: Wasn't working for some reason. Might have just been that I got the lodash-fp argument order wrong, but switching to .indexOf fixed it

Comment (zarsky-broad) [reply src/components/data/UploadPreviewTable.js:113]: Yup, we're using lodash's functional variant, so the array comes last.

File: src/components/data/FileBrowser.js
Comment (andrewmarcus) [reply src/components/data/FileBrowser.js]: Yeah, it's a nice trick to make time durations easier to read when you don't care about milliseconds. :)

File: src/components/data/FileBrowser.js
Comment (andrewmarcus) [reply src/components/data/FileBrowser.js:91]: Yep, I tried both and left some inconsistency in there. Sorry. Switched back to 5, since you're right, it provides the user with more immediate feedback.

**Comment by andrewmarcus** [comment]
Ok, posted some updates to resolve the things brought up in this PR. It also changes some more code in the Uploader codepath, so probably needs another glance. Sorry...

**Review by abaumann — APPROVED** [review]
Reviewed from a functional standpoint, and worked like a charm.  The UX for new vs updated columns is excellent - the color coding is clear and elegant

**Comment by ehanna4** [comment]
BTW looks like the integration tests won't work since this is a fork, so I ran them locally and they all passed 👍 Screen Shot 2021-04-06 at 10 26 26 AM

**Pull request merged**
